### PR TITLE
No delay when opening a link with notification enabled

### PR DESCRIPTION
**Pull request by Elihpmap**
Hello, I made this little fix after having tried this new version of BrowseRouter.
The issue was that the link was opening only after the notification had finished being displayed (around 5 to 10 seconds iirc). with this fix this shouldn't be the case anymore: both the notification and the link opening should happen simultaneously!
I also cleaned up the code according to what VisualStudio was flagging and I removed some VisualStudio config rules that weren't respected anyway. You can see the detail in the commits, (the actual fix is only the latest one).

I tested this on my end (on Windows 10) and everything seems to work well. I shouldn't have broken anything OS-wise as I only changed some internal logic and nothing that communicate with the OS. But tell me if you need more specific test before accepting this PR (I don't have any Win11 PC yet though)!

This is my first PR, but I have plans to make a few more on this project (which I really like by the way, thank you)! Please tell me if I did anything wrong that I need to correct here, or that I should do differently for potential future PR.

Have a good day!

### CODE REVIEW

**Comment by nref** [comment]
Thanks for your contribution, and for the great attitude!

Your most valuable contribution here is reporting the 10s delay bug. That is a great find, and very important to fix. I went ahead and released a fix in https://github.com/nref/BrowseRouter/releases/tag/0.12.5.

Is this your first PR ever, or do you mean only to BrowseRouter? There are some netiquette items to address. I don't have a contributing guide (yet), but here's what I would prefer

- Small PRs
- Separate PRs for features, bugfixes, and refactoring
- 1:1 ratio between issues and PRs

In short:

- Open an issue and PR for the 10s delay
- Open an issue and PR for the refactoring

Now to the contents of your PR. Since I already took care of the bugfix, I will only address the refactoring. I like many of the changes, but definitely can't accept some of them. I'll enumerate each in the comments. If you resolve them, I should be able to merge this PR, or if it's your preference, I'll also just make the changes and push that to your branch.

Thanks again!

File: .editorconfig [hunk 1]
```diff
@@ -60,14 +60,6 @@ dotnet_naming_rule.interface_names_must_begin_with_I.symbols   = interfaces
 dotnet_naming_rule.interface_names_must_begin_with_I.style = pascal_begin_with_I_style
 dotnet_naming_rule.interface_names_must_begin_with_I.severity = warning
 
-dotnet_naming_rule.private_and_protected_fields_must_begin_with_underscore.symbols   = private_fields
```
Comment (nref) [inline .editorconfig:63]: The editorconfig was correct - it's the code that is behind. I'd like private members to be `_likeThis` and public members to be `LikeThis`.

File: BrowseRouter/App.cs [hunk 2]
```diff
@@ -4,7 +4,7 @@ namespace BrowseRouter;
 
 public static class App
 {
-  private static string exePath;
```
Comment (nref) [inline BrowseRouter/App.cs:7]: Good catch on the nullable. Please format as `_exePath`

File: BrowseRouter/App.cs [hunk 3]
```diff
@@ -15,17 +15,17 @@ public static string ExePath
         return exePath;
       }
 
-      var assembly = Assembly.GetExecutingAssembly();
```
Comment (nref) [inline BrowseRouter/App.cs:18]: Good catch on the duplication.

File: BrowseRouter/BrowserService.cs [hunk 4]
```diff
@@ -4,7 +4,7 @@ namespace BrowseRouter;
 
 public class BrowserService(IConfigService config, INotifyService notifier)
 {
-  public async Task LaunchAsync(string url, string windowTitle)
```
Comment (nref) [inline BrowseRouter/BrowserService.cs:7]: This must remain async

File: BrowseRouter/BrowserService.cs [hunk 5]
```diff
@@ -38,7 +38,7 @@ public async Task LaunchAsync(string url, string windowTitle)
       Log.Write($"Launching {path} with args \"{args} {uri.OriginalString}\"");
 
       string name = GetAppName(path);
-      await notifier.NotifyAsync($"Opening {name}", $"URL: {url}");
```
Comment (nref) [inline BrowseRouter/BrowserService.cs:41]: The `await` keyword must remain. Removing it does let the current method continue as you intended, but then the program will exit before `NotifyAsync` has completed. On Windows 10, this will result in the tray icon remaining in the tray after the app has exited until the mouse moves overs over it, and then Windows removing the icon it doesn't respond to a window message. This is the issue adding the delay sought to fixed in the first place.

Moreover, in C# the correct way let the current method continue is by wrapping a call to `NotifyAsync` in `Task.Run`. As it currently is, any exception that `NotifyAsync` throws will be silently eaten, and on single-threaded environments this can deadlock (think Blazor).

File: BrowseRouter/ConfigService.cs [hunk 6]
```diff
@@ -8,18 +8,32 @@ public interface IConfigService
...
+  private IEnumerable<string> ConfigLines
   {
-    // Fix for self-contained publishing
-    this.ConfigPath = Path.Combine(Path.GetDirectoryName(App.ExePath)!, "config.ini");
+    get
+    {
+      if (configLines != null)
+        return configLines;
+
+      if (!File.Exists(ConfigPath))
```
Comment (nref) [inline BrowseRouter/ConfigService.cs:24]: I appreciate the added caching. Big getters are not good style. Could you please move the majority of this getter back into `ReadFile` ?

File: BrowseRouter/DefaultBrowserService.cs [hunk 7]
```diff
@@ -8,18 +8,18 @@ public class DefaultBrowserService(INotifyService notifier)
   private const string AppName = "BrowseRouter";
   private const string AppID = "BrowseRouter";
   private const string AppDescription = "Opens a different brower based on the URL";
-  private string AppIcon => App.ExePath + ",0";
-  private string AppOpenUrlCommand => App.ExePath + " %1";
+  private static string AppIcon => App.ExePath + ",0";
```
Comment (nref) [inline BrowseRouter/DefaultBrowserService.cs:11]: Please remove all of these static keywords. I understand your editor suggested these, but it does not know that just because something can be static does not mean it should be. 

Static is valuable where allocations are not wanted, such as in performance-critical code and hot paths. However, it's an enemy of maintainable code, because these members are not open for extension.

For example, keeping these nonstatic would keep open the possibility to expose them in an interface which `DefaultBrowserService` implements. Then they can be implemented in new ways, for example in test doubles. e.g. imagine `FakeBrowserService.GetRegisterStatus()`.

File: BrowseRouter/ConfigService.cs (hunk 6 above)
Comment (nref) [reply BrowseRouter/ConfigService.cs:24]: Strictly speaking, this is an added feature and belongs in a separate PR.

File: BrowseRouter/NotifyService.cs [hunk 8]
```diff
@@ -55,18 +55,18 @@ public async Task NotifyAsync(string title, string message)
...
     // On Windows 11, the pop-up remains for the full duration,
     // regardless if the app has exited or we delete the tray icon.
-    var delay = TimeSpan.FromSeconds(10);
+    TimeSpan delay = TimeSpan.FromSeconds(10);
     await Task.Delay(delay);
     Remove(nid);
   }
 
-  public void Remove(NotifyIconData nid)
+  public static void Remove(NotifyIconData nid)
```
Comment (nref) [inline BrowseRouter/NotifyService.cs:63]: Remove `static`

File: BrowseRouter/Program.cs [hunk 9]
```diff
@@ -35,7 +35,7 @@ private static async Task RunAsync(string arg)
       return;
     }
 
-    await LaunchUrlAsyc(arg);
+    LaunchUrl(arg);
```
Comment (nref) [inline BrowseRouter/Program.cs:38]: restore `await`

File: BrowseRouter/Program.cs [hunk 10]
```diff
@@ -61,22 +61,22 @@ private static async Task<bool> RunOption(string arg)
     return false;
   }
 
-  private static async Task LaunchUrlAsyc(string url)
+  private static void LaunchUrl(string url)
```
Comment (nref) [inline BrowseRouter/Program.cs:64]: restore `async`

File: BrowseRouter/Program.cs [hunk 11]
```diff
@@ -61,22 +61,22 @@ private static async Task<bool> RunOption(string arg)
...
-    var notifyPref = configService.GetNotifyPreference();
+    NotifyPreference notifyPref = configService.GetNotifyPreference();
     INotifyService notifier = notifyPref.IsEnabled switch
     {
       true => new NotifyService(),
       false => new EmptyNotifyService()
     };
 
-    await new BrowserService(configService, notifier).LaunchAsync(url, windowTitle);
+    new BrowserService(configService, notifier).Launch(url, windowTitle);
```
Comment (nref) [inline BrowseRouter/Program.cs:79]: restore `await`

**Review by nref — CHANGES_REQUESTED** [review]
See my comments.

File: BrowseRouter/DefaultBrowserService.cs (hunk 7 above)
Comment (Elihpmap) [reply BrowseRouter/DefaultBrowserService.cs:11]: I have trouble grasping how static could be an enemy of maintainable code, wouldn't it be revertable as soon as you need to make the variable instance-specific ?

Comment (Elihpmap) [reply BrowseRouter/DefaultBrowserService.cs:11]: Oh, no because that would mean a potentially bigger refactoring needed as you'd need to change how the no-longer-static variable is accessed everywhere else in the code! that's what you meant ?

Comment (nref) [reply BrowseRouter/DefaultBrowserService.cs:11]: Not necessarily if you've built other code that assumes it is static. The real problem here is static creates shared mutable state.

https://mb21.github.io/blog/2021/01/23/pure-functional-programming-and-shared-mutable-state

Static can be OK in the right context. This isn't it.

**Comment by nref** [comment]
> Oh Woaw! thanks for the detailed feedback! I'll answer your comment on which I have still some questions !
> 
> I believe this is my first PR ever yes, I've been leeching of github/gitlab for a while already though 😅
> 
> Since you already took care of the main bug, I think we can close this PR and I'll do the refactoring and the config issue&PR from the current master branch for a cleaner commit history (this shouldn't take me too long as most of it was done by my IDE anyway). Additionnaly for the refactor, I also tend to replace 'var' uses as I personnaly really dislike them. You didn't flag any in your comments but I'll understand if you prefer that I leave them as is too.
> 
> Thanks again for the detailed answer and comments!

I prefer `var` when it's clear on the right-hand side what the type is, otherwise explicit type. Thanks!

File: BrowseRouter/DefaultBrowserService.cs (hunk 7 above)
Comment (Elihpmap) [reply BrowseRouter/DefaultBrowserService.cs:11]: Thanks! I understand!
And theoretically if this variable was a const (it can't be in this case due to App.ExePath), then it would be a shared UNmutable state and it wouldn't be so bad, right ? I suppose that's why the preceding lines are fine being 'const' even if it would also disrupt code evolutions relying on them being non-static.

Comment (nref) [reply BrowseRouter/DefaultBrowserService.cs:11]: Yes, `const` makes it not mutable. So does `static readonly`, but `const` is special because it can be referenced at compile time.

**Comment by Elihpmap** [comment]
> I went ahead and released a fix in https://github.com/nref/BrowseRouter/releases/tag/0.12.5.

I think I miss-tested this release, the bug is still present and never actually disappeared...
I guess I still was using my wrong fix (which opened the link instantly but messed with the icon/notification display)

I opened a real issue for it this time, with a proposed fix in it: https://github.com/nref/BrowseRouter/issues/58
